### PR TITLE
Reduce warnings emitted by clang

### LINKED ISSUE #37: Clarity of -1 return for `unsigned char` functions
I assume all bits set is an error value. This corresponds a bit to `EOF` which is used by some C library functions for error result.
Typically such cases use `int` in as the return type. This may be less ideal on an 8-bit CPU when given the choice of return type.
Personally I think it is ugly to return -1 when the return type is unsigned. One can use `~0` or `~0u`, or perhaps better a macro to represent it.
Another way is to use `int8_t` as the return type which would signal that one can test for error by testing if the result is less than 0. In such case a negative value does not look inappropriate, which it kind of does now.

_Originally posted by @hth313 in https://github.com/MEGA65/mega65-libc/pull/29#discussion_r1267575969_

### PR DESCRIPTION
**Pull request by mlund**
This add the following changes to ongoing work on reducing the number of warnings emitted by clang:
- [x] Explicitly named stdint integers, e.g. `uint16_t` over `unsigned int` in `memory.h`
- [x] Memory 28-bit addresses changed from signed `long` to unsigned `uint32_t` in `memory.h`
- [x] Explicitly pack structs on clang (for llvm-mos-sdk version v2.0.0 or earlier, see fix in https://github.com/llvm-mos/llvm-mos/commit/09aa314634ec1a3b9c56fdf8f19fc8bd71fa547d)
- [x] Add test-integer test target
- [x] Slightly stronger test of `dma_poke()` and `dma_peek()`
- [x] Removed redundant includes on cc65
- [x] Return named error instead of -1 in sdcard (see #37).
- [x] Partial replacement of implicit casts with explicit ones

Closes #37.

### CODE REVIEW

File: src/memory.c
Comment (hth313) [inline src/memory.c:58]: These variables are used locally in the following function, why not put then inside it? Being outside like this their symbols have global visibility.
Look at the code, the do not even need to be static, it seems they can be ordinary auto variables?

File: src/random.c
Comment (hth313) [inline src/random.c:27]: These `random` variables have global scope, maybe they should be `static`?

File: src/sdcard.c
Comment (hth313) [inline src/sdcard.c:11]: This probably works, but I wonder why `U` is used, to silence a warning? Maybe try without the `U` or use `UL` as technically it is a 16 bit unsigned constant (without the `L`).

File: src/sdcard.c
Comment (hth313) [inline src/sdcard.c:306]: Nitpicking, but if you find that you need casts like this, it is often an indication that you are using inconsistent (signed/unsigned) types. In a perfect world, any needed casts in situations of low memory or I/O is best done in the lowest routine that needs to deal with the actual I/O and layers above should use consistent types.
But yeah, it may be beyond the scope of this PR.

Comment (mlund) [reply src/sdcard.c:306]: I fully agree that we shouldn't need to cast, but wanted to at least make the mismatch explicit in the code as e.g. cc65 doesn't warn about this. This should be fixed more long term, but I would wait until we have more tests.

File: src/memory.c
Comment (mlund) [reply src/memory.c:58]: Makes sense to me and I don't know the original motivation. There might be different optimization strategies for cc65(?). Perhaps open a separate PR focusing on random?

File: src/sdcard.c
Comment (mlund) [reply src/sdcard.c:11]: Here the main change is from signed to unsigned 32 bit, so it should be UL. Will fix that. In any case there's no loss in precision so no complaints from the compiler.

File: src/random.c
Comment (mlund) [reply src/random.c:27]: Perhaps in a separate PR, see above?

File: src/fat32.c
Comment (mlund) [reply src/fat32.c:31]: Yes I agree and my main aim to make these casts explicit as cc65 doesn't want about these (see also comment below). There's no test of fat32.c so I don't dare to change this before that's in place.

File: src/memory.c
Comment (hth313) [reply src/memory.c:58]: It may be because there is the idea that you should avoid local variables on some 6502 compilers, so perhaps it is best to do it conditionally depending on the compiler. It is probably best to leave it for now.

File: src/sdcard.c
Comment (hth313) [reply src/sdcard.c:306]: It sounds like a good idea to wait until there are more tests in place.

File: src/random.c
Comment (mlund) [reply src/random.c:27]: @hth313 If there's nothing more for this PR, would you please click approve? (We now require at least one approved review before merging PR's)

**Review by hth313 — APPROVED** [review]

File: src/random.c
Comment (hth313) [reply src/random.c:27]: I have approved, but I do not think it helps as I do not have commit rights on this repository.

Comment (mlund) [reply src/random.c:27]: Thanks! Let's ping @lydon42 for help.

Comment (lydon42) [reply src/random.c:27]: Random could be replaced soon by neoTRNG "real" random...

File: src/fat32.c
Comment (lydon42) [inline src/fat32.c:31]: why not uint16_t?

File: src/fat32.c
Comment (lydon42) [inline src/fat32.c:141]: why no stdint type?

**Review by lydon42 — COMMENTED** [review]
Is there a reason for the two changes I pointed out not to change to a stdint type?

File: src/fat32.c
Comment (mlund) [reply src/fat32.c:31]: Because `clusters` is still `unsigned short`. I agree that stdint types would be better, but I'd prefer to refactor this in a separate PR, and preferable when a test is in place. Here the idea is to convert implicit casts to explicit ones.

File: src/fat32.c
Comment (mlund) [reply src/fat32.c:141]: Because `root_dir_sector` is still `long`. I agree that stdint types would be better, but I'd prefer to refactor this in a separate PR, and preferable when a test is in place. Here the idea is to convert implicit casts to explicit ones.

**Review by lydon42 — APPROVED** [review]
Ok, sounds good to me!

**Pull request merged**
